chat-client.py

```diff
@@ -0,0 +1,64 @@
+import socket
+import selectors
+import errno
+import sys
+import time
+
+HEADER_LENGTH = 10
+HOST = '127.0.0.1'
+PORT = 54321
+
+print('\nWelcome to the chat server!\n')
+time.sleep(1)
+print('Hit enter to send message or refresh message feed\n')
+time.sleep(1)
+print('Type "quit" to leave the chat\n')
+time.sleep(1)
+my_username = input('Enter username: ')
+
+# Set up client socket
+sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+sock.connect((HOST, PORT))
+sock.setblocking(False)
+
+# Send username to server with a header
+my_username_enc = my_username.encode("utf-8")
+my_username_header = f'{len(my_username_enc):<{HEADER_LENGTH}}'.encode("utf-8")
+sock.send(my_username_header + my_username_enc)
+
+while True:
+	msg = input(f'{my_username}> ')
+	if msg.lower() == 'quit':
+		sock.close()
+		print('Connection closed')
+		sys.exit()
+	elif msg:
+		# Send message
+		msg_enc = msg.encode("utf-8")
+		msg_header = f'{len(msg_enc):<{HEADER_LENGTH}}'.encode("utf-8")
+		sock.send(msg_header + msg_enc)
+	try:
+		# Receive messages
+		while True:
+			user_header = sock.recv(HEADER_LENGTH)
+			if not user_header:
+				print('Connection closed by server')
+				sys.exit()
+			user_len = int(user_header.decode("utf-8").strip())
+			user = sock.recv(user_len).decode("utf-8")
+			msg_header = sock.recv(HEADER_LENGTH)
+			msg_len = int(msg_header.decode("utf-8").strip())
+			msg = sock.recv(msg_len).decode("utf-8")
+			print(f'{user}> {msg}')
+	except IOError as e:
```

Comment:
Don't use exceptions for flow control. Here the fact an exception is being thrown and caught allows you to detect that there is no more data to be read, i.e. this is controlling your program flow. Exceptions should only be used when there is a genuine error. The reason to try and avoid this is because exceptions are actually expensive resource-wise.

I think the solution here would again to use selectors to detect when there is data ready to be read, and if you still get this exception you know there's a genuine problem

Reply:
Ok I will have a look at this after you've responded to this first set of questions. One potential issue I see with using selectors is that selectors.select() blocks until there's a socket ready for I/O. So won't this cause an issue with the user being unable to type a message once the loop hits this point?

Reply:
Ok I will have a look at this after you've responded to this first set of questions. One potential issue I see with using selectors is that selectors.select() blocks until there's a socket ready for I/O. So won't this cause an issue with the user being unable to type a message once the loop hits this point?

Reply:
Parked this for now. It worked but there was in issue with received messages interrupting input()